Implement SEP-990 Enterprise Managed OAuth

BinoyOza-okta opened the pull request:
This PR implements the **client-side** components of **SEP-990: Enterprise Managed Authorization**. It introduces the `EnterpriseAuthOAuthClientProvider` to handle the full token exchange flow required for Enterprise SSO, including RFC 8693 (Token Exchange) and RFC 7523 (JWT Bearer Grant).

## Motivation and Context
**Implements:** [SEP-990](https://github.com/modelcontextprotocol/modelcontextprotocol/issues/990) #1593 

To support enterprise environments where direct API keys are not compliant, the Python SDK needs to support "Managed Authorization." This implementation allows the SDK to:
1.  **Exchange Tokens:** Convert an existing Identity Provider (IdP) token (e.g., OIDC ID Token or SAML Assertion) into an MCP-specific "ID-JAG" token using **RFC 8693**.
2.  **Authenticate:** Exchange the ID-JAG token for a usable Access Token via **RFC 7523**.

This aligns the Python SDK with the architecture defined in the SEP-990 implementation guide.

## Implementation Details
The following components have been added to `src/mcp/client/auth/extensions/enterprise_managed_auth.py`:
* **Data Models:** `TokenExchangeParameters` and `TokenExchangeResponse` using Pydantic to strictly type the exchange payloads.
* **Client Provider:** `EnterpriseAuthOAuthClientProvider`, which extends the base `OAuthClientProvider` to orchestrate the exchange logic.
* **ID-JAG Logic:** Support for `urn:ietf:params:oauth:token-type:id-jag` token types.

## How Has This Been Tested?
I have implemented comprehensive unit tests in `tests/client/auth/test_enterprise_managed_auth_client.py` using `pytest` and `unittest.mock`.

The testing suite covers the following scenarios:

1.  **Data Model Validation:**
    * Verified `TokenExchangeParameters` correctly generates requests for both **OIDC ID Tokens** (`test_token_exchange_params_from_id_token`) and **SAML Assertions** (`test_token_exchange_params_from_saml_assertion`).
    * Validated enforcement of required fields (audience, resource, subject_token).

2.  **RFC 8693 Token Exchange Logic:**
    * **Success Path:** Mocked `httpx` to verify the correct payload structure (grant types, token types) is sent to the IdP.
    * **Client Authentication:** Verified that `client_id` and `client_secret` are correctly injected into the request body when configured (`test_exchange_token_with_client_authentication`).
    * **Error Handling:** Validated graceful handling of HTTP 400/500 errors, non-JSON error responses, and unexpected token types.

3.  **RFC 7523 JWT Bearer Grant Logic:**
    * **Success Path:** Verified the exchange of an ID-JAG token for a final Access Token.
    * **Metadata Checks:** Ensured the flow fails appropriately if the MCP server's OAuth metadata (token endpoint) is missing.

4.  **Network Edge Cases:**
    * Simulated network failures (`httpx.ConnectError`, `httpx.ReadTimeout`) to ensure `OAuthTokenError` is raised with descriptive messages.
## Breaking Changes
**No.**
This is an additive extension. The core `OAuthClientProvider` remains backward compatible. Only users specifically importing and using `EnterpriseAuthOAuthClientProvider` will be affected.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
* **Dependencies:** Utilizes `pydantic` for model validation and `httpx` for async requests.
* **Scope:** As agreed, this PR includes the **Client** implementation (`src/mcp/client/auth/extensions/`).

BinoyOza-okta committed:
- Implemented SEP-990 feature for providing support for Enterprise Managed Auth support.
- Written unit test cases for client and server implementation of the enterprise managed auth code.

BinoyOza-okta committed:
Added test cases for missing lines of code.

BinoyOza-okta committed:
- Added tests cases for few of the missing lines. src/mcp/client/auth/extensions/enterprise_managed_auth.py 232->235, 304->307.
- Resolved pre-commit errors.

BinoyOza-okta committed:
- Fixed pre-commit errors.

BinoyOza-okta committed:
- Tried to fix the ruff error.

BinoyOza-okta committed:
- Fixed ruff errors.

BinoyOza-okta committed:
- Removed server side changes for enterprise_managed_auth.py

BinoyOza-okta committed:
- Added README.md changes for SEP-990 implementation for enterprise managed auth.

BinoyOza-okta committed:
- Resolved pyright checks error.

BinoyOza-okta committed:
- Resolved README.md file fixes for removing unused imports.

BinoyOza-okta committed:
- Resolved pyright errors.

BinoyOza-okta committed:
- Added new test cases for the missing code lines.

BinoyOza-okta committed:
- Fixed the failing test cases.

BinoyOza-okta committed:
- Fixed the test cases.

maxisbey commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,412 @@
+    Note:
+        For verification, use server-side validation instead.
+    """
+    import jwt
```
move to top of file

maxisbey commented on README.md:
````diff
@@ -2452,6 +2357,129 @@ _Full example: [examples/snippets/clients/oauth_client.py](https://github.com/mo
+
+**Example Usage:**
+
+```python
````
Can this instead be put in the snippets folder and then use the `scripts/update_readme_snippets.py` script to update the README accordingly.

maxisbey reviewed (changes requested):
Unless I was missing something,

BinoyOza-okta committed:
- Added typing for request payload structures TokenExchangeRequestData and JWTBearerGrantRequestData.
- Added snippet file for adding code to the README.md file.
- Added new section in README.md file to add information regarding: "how to use the access token once you get it" and "How does this work when the client ID is expired?".

BinoyOza-okta commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,412 @@
+        idp_token_endpoint: str,
+        token_exchange_params: TokenExchangeParameters,
+        redirect_handler: Any = None,
+        callback_handler: Any = None,
```
Mark this with correct typing.

BinoyOza-okta committed:
- Updated test case to include IDJAGClaims type model to verify payload.

BinoyOza-okta commented on tests/client/auth/test_enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,1126 @@
+
+# ============================================================================
+# Tests for TokenExchangeParameters
+# ============================================================================
```
Resolved

BinoyOza-okta commented:
Hi @maxisbey, I have addressed all your comments. Could you please review the PR?

BinoyOza-okta committed:
feat: Add conformance tests for enterprise managed authorization (SEP-990)

Implements comprehensive conformance testing for enterprise managed
authorization flows including RFC 8693 token exchange and RFC 7523
JWT bearer grant with ID-JAG tokens.

- Add 3 conformance test scenarios to client.py:
  * auth/enterprise-id-jag-validation - Validates ID-JAG token structure
  * auth/enterprise-token-exchange - Tests OIDC ID Token → ID-JAG → Access Token flow
  * auth/enterprise-saml-exchange - Tests SAML Assertion → ID-JAG → Access Token flow

- Create enterprise_auth_server.py:
  * Implements RFC 8693 token exchange endpoint (/token-exchange)
  * Implements RFC 7523 JWT bearer grant endpoint (/oauth/token)
  * Provides OAuth metadata endpoint for discovery
  * Supports both OIDC ID tokens and SAML assertions
  * Issues ID-JAG tokens with proper structure (typ: oauth-id-jag+jwt)
  * Validates bearer tokens and provides protected MCP endpoints

- Add run-enterprise-auth-with-server.sh:
  * Starts mock server on port 3002
  * Dynamically fetches test context
  * Runs all 3 enterprise auth scenarios
  * Reports detailed test results
  * Cleans up servers on exit

- Update conformance.yml workflow:
  * Add enterprise-auth-conformance job
  * Runs on every pull request
  * Marked as optional (continue-on-error: true)
  * Tests run in parallel with other conformance checks

- Add fastapi>=0.115.0 to dev dependencies
  * Required for mock server implementation
  * Only needed for conformance testing
  * Update uv.lock accordingly

- Fix docstring formatting in enterprise_managed_auth_client.py:
  * Update get_id_token_from_idp() to follow PEP 257 (D212)
  * Fix multi-line docstring to start summary on first line
  * Apply fix to both example file and README.md
  * Ensures all example tests pass

- Minor updates to enterprise_managed_auth.py:
  * Improve error handling
  * Add validation for token exchange parameters

✅ ID-JAG Token Validation
✅ OIDC ID Token Exchange Flow
✅ SAML Assertion Exchange Flow

- RFC 8693 Token Exchange (ID Token and SAML)
- RFC 7523 JWT Bearer Grant
- ID-JAG token structure validation
- OAuth metadata discovery
- Bearer token authentication
- Error handling and edge cases

- SEP-990: Enterprise Managed Authorization
- RFC 8693: OAuth 2.0 Token Exchange
- RFC 7523: JWT Profile for OAuth 2.0 Client Authentication
- RFC 8707: Resource Indicators for OAuth 2.0

Run conformance tests:
```bash
./.github/actions/conformance/run-enterprise-auth-with-server.sh

# Conflicts:
#	.github/workflows/conformance.yml

BinoyOza-okta committed:
- Fix uv.lock.

pcarleton commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,204 @@
+    EnterpriseAuthOAuthClientProvider,
+    TokenExchangeParameters,
+)
+from mcp.client.sse import sse_client
```
please use streamablehttp_client

pcarleton commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,204 @@
+        # Step 7: Use the access token to connect to MCP server
+        headers = {"Authorization": f"Bearer {access_token.access_token}"}
+
+        async with sse_client(url="https://mcp-server.example.com", headers=headers) as (read, write):
```
please use streamablehttp_client

pcarleton commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,204 @@
+    # Step 2: Configure token exchange parameters
+    token_exchange_params = TokenExchangeParameters.from_id_token(
+        id_token=id_token,
+        mcp_server_auth_issuer="https://your-idp.com",  # IdP issuer URL
```
this should not be the IDP URL, it's the issuer for the MCP server.  so auth.mcp.example.com maybe

pcarleton commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,416 @@
+            "assertion": id_jag,
+        }
+
+        # Add client authentication
```
I believe you'll want to follow what the AS says it supports here, so either client_secret_post or client_secret_basic

BinoyOza-okta replied:
Updated to have client_secret_basic

pcarleton commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,416 @@
+        client_metadata: OAuthClientMetadata,
+        storage: TokenStorage,
+        idp_token_endpoint: str,
+        token_exchange_params: TokenExchangeParameters,
```
i think it's a bit awkward to take in TokenExchangeParameters here. 

I think `id_token_provider: Callable[[], Awaitable[str]],` would be a good argument. 

we also likely need `idp_client_id` / `idp_client_secret` as arguments.

I think it'd be good to rely on the existing mechanisms to get the resource URL and AS metadata, e.g.:

```
  async def _perform_authorization(self) -> httpx.Request:
      audience = str(self.context.oauth_metadata.issuer)
      resource = self.context.get_resource_url()
      id_token = await self.id_token_provider()

      # Side request to IdP for token exchange (RFC 8693)
      id_jag = await self._exchange_for_id_jag(id_token, audience, resource)

      # Return JWT bearer grant request (RFC 7523) for base class to send
      return self._build_jwt_bearer_request(id_jag)
```

BinoyOza-okta committed:
feat(auth): migrate enterprise auth conformance tests to @modelcontextprotocol/conformance v0.1.14

Refactored enterprise managed authorization implementation and conformance tests
to use the official @modelcontextprotocol/conformance npm package (v0.1.14)
instead of custom mock servers.

**Key Changes:**

**Implementation:**
- Refactored `EnterpriseAuthOAuthClientProvider._perform_authorization()` to
  return `httpx.Request` instead of executing requests directly
- Moved error handling to parent class `OAuthClientProvider.async_auth_flow()`
- Updated method signatures:
  - `exchange_token_for_id_jag(client)` → returns `str` (ID-JAG)
  - `exchange_id_jag_for_access_token(id_jag)` → returns `httpx.Request` (not OAuthToken)

**Conformance Tests:**
- Migrated from custom mock server to official conformance package v0.1.14
- Removed custom `enterprise_auth_server.py` (332 lines)
- Removed custom `run-enterprise-auth-with-server.sh` (169 lines)
- Added `run-enterprise-auth-conformance.sh` using official conformance scenarios
- Updated `client.py` to support SEP-990 conformance tests:
  - `auth/cross-app-access-complete-flow`
  - `auth/enterprise-token-exchange`
  - `auth/enterprise-jwt-bearer`
- All 9/9 conformance checks passing

**Tests:**
- Added 2 new test cases for 100% code coverage
- Updated 32 existing tests to match new implementation
- Removed 3 duplicate/skipped tests
- Total: 29 tests passing, 0 failing, 100% coverage

**Documentation:**
- Updated `README.md` Enterprise Managed Authorization section
- Replaced manual token exchange examples with automatic auth flow pattern
- Added advanced manual flow example showing correct method signatures
- Updated `examples/snippets/clients/enterprise_managed_auth_client.py`
- Removed unused imports

**Related:**
- PR: https://github.com/modelcontextprotocol/conformance/pull/110
- Spec: SEP-990 (Enterprise Managed Authorization)
- Package: @modelcontextprotocol/conformance@0.1.14

**Testing:**
- ✅ All unit tests passing
- ✅ 100% code coverage for enterprise_managed_auth.py
- ✅ Conformance tests passing (9/9 checks)
- ✅ GitHub Actions workflow updated

BinoyOza-okta committed:
Fixed end-of-file-fixer for test_enterprise_managed_auth_client.py.

BinoyOza-okta committed:
Fixed pre-commit hooks issues.

BinoyOza-okta committed:
Fixed Ruff formatting issues.

BinoyOza-okta committed:
Fixed pyright issues.

BinoyOza-okta committed:
Moved the changes from README.md to README.v2.md.

BinoyOza-okta committed:
- Removed Unicode characters from the example snippet file.
- Removed FastAPI dependency as it is no longer required.

BinoyOza-okta commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,204 @@
+                await session.initialize()
+
+                # Call tools with automatic retry on token expiration
+                result = await call_tool_with_retry(
```
Have updated the code snippets and moved them from README.md to README.v2.md.

BinoyOza-okta committed:
- Removed unused parameters from the EnterpriseAuthOAuthClientProvider class.
- Updated docstring for decode_id_jag method.

pcarleton commented on .github/actions/conformance/client.py:
```diff
@@ -314,6 +326,266 @@ async def run_auth_code_client(server_url: str) -> None:
+    await _run_auth_session(server_url, enterprise_auth)
+
+
+@register("auth/enterprise-token-exchange")
```
please delete these and below

BinoyOza-okta replied:
Have deleted the non-existent scenarios.

BinoyOza-okta committed:
refactor(auth): address code review feedback for enterprise managed auth

Addresses code review comments for SEP-990 enterprise auth implementation:

1. Removed non-existing scenarios from client.py.
2. Removed run-enterprise-auth-conformance.sh script.
3. Remove redundant CI job, use conformance v0.1.14 with --suite all
4. Document specific use cases for manual token exchange flow
5. Added client_credentials support to the __init__.py
6. Rename TokenExchangeResponse to IDJAGTokenExchangeResponse for clarity
7. Fix IdP credential usage - separate idp_client_id/secret from MCP credentials
8. Add parameter discovery example demonstrating OAuth metadata discovery

BinoyOza-okta committed:
- Removed the non-existing scenario from docstring.

BinoyOza-okta committed:
### Code Review Fixes

**Round 1 - Initial Review:**
- ✅ Added ID-JAG expiry tracking to prevent stale tokens
- ✅ Enhanced audience override with WARNING level logging
- ✅ Documented error handling delegation pattern
- ✅ Added metadata stability documentation
- ✅ Fixed empty scope handling with strip() check
- ✅ Added negative JWT decode tests

**Round 2 - Systematic Review:**
- ✅ Made ID-JAG expiry configurable (DEFAULT_ID_JAG_EXPIRY_SECONDS = 900)
- ✅ Added comprehensive thread safety documentation.
- ✅ Moved time import to module level
- ✅ Extracted magic numbers to constants
- ✅ Enhanced logging with default expiry notifications

**Round 3 - Final Fixes:**
- ✅ Changed validation exceptions to OAuthFlowError for consistency
- ✅ Added edge case test for oauth_metadata=None
- ✅ Fixed critical bug: instance variable vs class constant

### Test Coverage

**Statistics:**
- 43 test cases covering all code paths
- 100% code coverage (166 statements, 44 branches)
- Edge cases: expired tokens, malformed JWT, HTTP errors, empty scope
- Conformance test integration: `auth/cross-app-access-complete-flow`

### Breaking Changes

**None.** All changes are backward compatible with optional parameters.

BinoyOza-okta committed:
- Fixed end of file for test_enterprise_managed_auth_client.py.

BinoyOza-okta committed:
- Remove dead is_token_valid() guard from _perform_authorization; parent only calls this when new tokens are needed, so it runs unconditionally now
- Replace JWTBearerGrantRequestData TypedDict + double-cast with plain dict[str, str] matching prepare_token_auth signature
- Stop pre-adding client_secret to POST body in JWT bearer grant; prepare_token_auth already handles placement per auth method
- Add override_audience_with_issuer constructor param (default True) so callers in federated/multi-tenant setups can opt out of audience override
- Stop mutating injected TokenExchangeParameters; use internal _subject_token field for refresh_with_new_id_token
- Call validate_token_exchange_params in __init__ for fail-fast validation
- Raise ValueError for idp_client_secret without idp_client_id
- Use str.removesuffix("/mcp") in conformance client instead of str.replace

BinoyOza-okta committed:
fix(auth): harden enterprise managed auth with RFC compliance fixes

Fix several correctness issues in EnterpriseAuthOAuthClientProvider
found during post-review audit:

- Add missing resource (RFC 8707) and scope params to JWT bearer grant,
  matching the pattern used by all sibling providers
- Move constructor validation before super().__init__() to avoid
  orphaned OAuthContext on validation failure
- Catch pydantic ValidationError in exchange_token_for_id_jag when IdP
  returns HTTP 200 with malformed JSON
- Wrap refresh_with_new_id_token in context lock to prevent racing with
  concurrent async_auth_flow
- Remove write-only dead state (_id_jag, _id_jag_expiry,
  default_id_jag_expiry, DEFAULT_ID_JAG_EXPIRY_SECONDS)
- Use set literal instead of list for membership test in
  validate_token_exchange_params
- Delegate client auth entirely to prepare_token_auth() instead of
  manually injecting client_id and mutating token_endpoint_auth_method
- Use _get_token_endpoint() instead of requiring oauth_metadata
- Improve docstrings for IDJAGClaims, validate_token_exchange_params,
  and concurrency notes

BinoyOza-okta commented on .github/actions/conformance/client.py:
```diff
@@ -16,7 +16,19 @@
+    auth/enterprise-token-exchange          - Enterprise auth with OIDC ID token (legacy name)
+    auth/enterprise-saml-exchange           - Enterprise auth with SAML assertion (legacy name)
+    auth/enterprise-id-jag-validation       - Validate ID-JAG token structure (legacy name)
     auth/*                                  - Authorization code flow (default for auth scenarios)
```
Removed the scenarios that do not exist, and kept only the top one.

BinoyOza-okta commented on src/mcp/client/auth/extensions/__init__.py:
```diff
@@ -0,0 +1,19 @@
+"""MCP Client Auth Extensions."""
+
+from mcp.client.auth.extensions.enterprise_managed_auth import (
```
Added for both of them

BinoyOza-okta commented on .github/actions/conformance/run-enterprise-auth-conformance.sh:
```diff
@@ -0,0 +1,89 @@
+#!/bin/bash
```
Removed the script.

BinoyOza-okta commented on .github/workflows/conformance.yml:
```diff
@@ -43,3 +43,18 @@ jobs:
       - run: uv sync --frozen --all-extras --package mcp
       - run: npx @modelcontextprotocol/conformance@0.1.13 client --command 'uv run --frozen python .github/actions/conformance/client.py' --suite all
+
+  enterprise-auth-conformance:
```
Have updated the workflow not to run the enterprise-managed-auth conformance test as a separate test. Now --suite all is running all the tests.

BinoyOza-okta commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,146 @@
+
+
+async def advanced_manual_flow() -> None:
+    """Advanced example showing manual token exchange (for special use cases)."""
```
Have updated the docstring to describe specific use cases rather than the special-use-case comment.

BinoyOza-okta commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,416 @@
+        )
+
+
+class TokenExchangeResponse(BaseModel):
```
Thanks, have updated the model name.

BinoyOza-okta commented on src/mcp/client/auth/extensions/enterprise_managed_auth.py:
```diff
@@ -0,0 +1,416 @@
+
+        # Add client authentication if needed
+        if self.context.client_info:
+            if self.context.client_info.client_id is not None:
```
Yes, correct, I have addressed and corrected it.

BinoyOza-okta commented on examples/snippets/clients/enterprise_managed_auth_client.py:
```diff
@@ -0,0 +1,146 @@
+    token_exchange_params = TokenExchangeParameters.from_id_token(
+        id_token=id_token,
+        mcp_server_auth_issuer="https://auth.mcp-server.example.com",  # MCP server's auth issuer
+        mcp_server_resource_id="https://mcp-server.example.com",  # MCP server resource ID
```
I have updated the snippet to fetch the parameter rather than hardcode it.

km-anthropic commented:
@claude review

nima54851 commented:
Solid contribution! The structured output implementation looks clean.

Quick feedback: the error handling in memory.py could benefit from more granular retry logic for transient failures. Also, has the team considered adding structured logging from the start? It'll save a lot of debugging pain later.

Happy to help if you want a hand with any of this!
